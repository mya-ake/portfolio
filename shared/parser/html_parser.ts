import { DefaultTreeAdapterMap, parse } from "parse5";

type NodeType = "tag" | "text";
type Attributes = Record<string, string>;
type Document = DefaultTreeAdapterMap["document"];
type ChildNode = Document["childNodes"][number];
type Element = DefaultTreeAdapterMap["element"];
type Parse5TextNode = DefaultTreeAdapterMap["textNode"];

export type TagNode = {
  nodeType: "tag";
  tagName: string;
  attrs: Attributes;
  childNodes: Node[];
};

export type TextNode = {
  nodeType: "text";
  content: string;
};

export type Node = TagNode | TextNode;

const getNodeType = (nodeName: string): NodeType => {
  switch (nodeName) {
    case "#text":
      return "text";
    default:
      return "tag";
  }
};

const createTagNode = (childNode: Element): TagNode => ({
  nodeType: "tag",
  tagName: childNode.tagName,
  attrs: childNode.attrs.reduce<Attributes>((attrs, attrItem) => {
    attrs[attrItem.name] = attrItem.value;
    return attrs;
  }, {}),
  childNodes: convertNodes(childNode.childNodes ?? []),
});

const createTextNode = (childNode: Parse5TextNode): TextNode => ({
  nodeType: "text",
  content: childNode.value,
});

const convertNode = (childNode: ChildNode): Node => {
  const nodeType = getNodeType(childNode.nodeName);
  switch (nodeType) {
    case "tag":
      return createTagNode(childNode as Element);
    case "text":
      return createTextNode(childNode as Parse5TextNode);
  }
};

const convertNodes = (childNodes: ChildNode[]): Node[] =>
  childNodes.map(convertNode);

const extractBaseNodes = (nodes: Node[]): Node[] => {
  return nodes.reduce((currentNodes, node) => {
    if (node.nodeType === "tag") {
      if (["html", "body"].includes(node.tagName)) {
        return extractBaseNodes(node.childNodes);
      }
    }
    return [...currentNodes, node];
  }, [] as Node[]);
};

export type Preprocessor = (html: string) => string;
export type Option = {
  preprocessors?: Preprocessor[];
};

export const parseHtml = (html: string, option?: Option): Node[] => {
  const preprocessors = option?.preprocessors ?? [];
  const processedHtml = preprocessors.reduce(
    (html, processor) => {
      return processor(html);
    },
    html,
  );
  const documentFragment = parse(processedHtml, {});
  const nodes = convertNodes(documentFragment.childNodes);
  return extractBaseNodes(nodes);
};
