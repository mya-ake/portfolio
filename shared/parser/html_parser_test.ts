import { assertEquals } from "std/testing/asserts.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";
import { parseHtml } from "./html_parser.ts";

/**
 * Single root HTML
 */
Deno.test({
  name: "parse html without attrs",
  fn() {
    const result = parseHtml(`<p>text</p>`);
    assertEquals(result, [
      {
        nodeType: "tag",
        tagName: "p",
        attrs: {},
        childNodes: [
          {
            nodeType: "text",
            content: "text",
          },
        ],
      },
    ]);
  },
});

Deno.test({
  name: "parse html with attrs",
  fn() {
    const result = parseHtml(
      `<button type="button" disabled aria-pressed="false">text</button>`
    );
    assertEquals(result, [
      {
        nodeType: "tag",
        tagName: "button",
        attrs: {
          type: "button",
          disabled: "",
          "aria-pressed": "false",
        },
        childNodes: [
          {
            nodeType: "text",
            content: "text",
          },
        ],
      },
    ]);
  },
});

/**
 * Multi root HTML
 */
Deno.test({
  name: "parse multi rool html",
  fn() {
    const result = parseHtml(
      `<p class="p p1">text 1</p><p class="p">text 2</p>`
    );
    assertEquals(result, [
      {
        nodeType: "tag",
        tagName: "p",
        attrs: {
          class: "p p1",
        },
        childNodes: [
          {
            nodeType: "text",
            content: "text 1",
          },
        ],
      },
      {
        nodeType: "tag",
        tagName: "p",
        attrs: {
          class: "p",
        },
        childNodes: [
          {
            nodeType: "text",
            content: "text 2",
          },
        ],
      },
    ]);
  },
});

/**
 * deep children
 */
Deno.test({
  name: "parse list",
  async fn(t) {
    const result = parseHtml(`<ul><li>item 1</li><li>item 1</li></ul>`);
    await assertSnapshot(t, result);
  },
});

Deno.test({
  name: "parse some content",
  async fn(t) {
    const result = parseHtml(`<div>no wrapped<p>item 1</p><img src="/static/sample.png" alt=""/></div>`);
    await assertSnapshot(t, result);
  },
});

/**
 * web components
 */
 Deno.test({
  name: "parse some content",
  async fn(t) {
    const result = parseHtml(`<custom-p>text</custom-p>`);
    await assertSnapshot(t, result);
  },
});
