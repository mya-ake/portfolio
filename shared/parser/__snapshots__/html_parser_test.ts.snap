export const snapshot = {};

snapshot[`parse list 1`] = `
[
  {
    attrs: {},
    childNodes: [
      {
        attrs: {},
        childNodes: [
          {
            content: "item 1",
            nodeType: "text",
          },
        ],
        nodeType: "tag",
        tagName: "li",
      },
      {
        attrs: {},
        childNodes: [
          {
            content: "item 1",
            nodeType: "text",
          },
        ],
        nodeType: "tag",
        tagName: "li",
      },
    ],
    nodeType: "tag",
    tagName: "ul",
  },
]
`;

snapshot[`parse some content 1`] = `
[
  {
    attrs: {},
    childNodes: [
      {
        content: "no wrapped",
        nodeType: "text",
      },
      {
        attrs: {},
        childNodes: [
          {
            content: "item 1",
            nodeType: "text",
          },
        ],
        nodeType: "tag",
        tagName: "p",
      },
      {
        attrs: {
          alt: "",
          src: "/static/sample.png",
        },
        childNodes: [
        ],
        nodeType: "tag",
        tagName: "img",
      },
    ],
    nodeType: "tag",
    tagName: "div",
  },
]
`;

snapshot[`parse some content 2`] = `
[
  {
    attrs: {},
    childNodes: [
      {
        content: "text",
        nodeType: "text",
      },
    ],
    nodeType: "tag",
    tagName: "custom-p",
  },
]
`;
